### PR TITLE
Add documentation to DG for `undo`, `redo` and `history`

### PR DESCRIPTION
**Pull request by vigneshsankariyer1234567890**
Details the implementation of the various classes which allow the history of states to be stored.

Here are some of the main diagrams which are being used:
HistoryCommandSequenceDiagram
HistorySequenceToStoreCommandsDiagram
ModelClassDiagram
UndoSequenceDiagram
DownButtonActivityDiagram
UpButtonActivityDiagram

### CODE REVIEW

File: docs/DeveloperGuide.md [hunk 1]
```diff
@@ -351,6 +346,161 @@ it will be useful for them to interact with their contact lists through the cust
...
+
+Step 3. The user changes the current state of Teletubbies by executing a `delete -i 1` command. The `delete` command calls `Model#commitAddressBook()` as before, causing 
+the state of Teletubbies to be changed.
+![UndoState2](images/UndoRedoState2.png)
+
+Step 4. The user decides to undo by executing `undo`. This causes `Model#undoAddressBook()` to be called, which causes `HistoryManager` to revert back to state `tb1`.
+![UndoState3](images/UndoRedoState3.png)
+
+<div markdown="span" class="alert alert-info">:information_source: **Note:** If the current state is the earliest 
+possible state, in this case tb0, Teletubbies will no longer be revertible, and the undo command will throw an exception. The undo command uses
```
Comment (skythefire) [inline docs/DeveloperGuide.md:446]: I think there is a typo here!

File: docs/DeveloperGuide.md [hunk 2]
```diff
@@ -351,6 +346,161 @@ it will be useful for them to interact with their contact lists through the cust
...
+`Model#canUndoAddressBook()` to confirm if this is the case.
+</div>
+
+The sequence diagram below illustrates the undo operation:
+![UndoSequenceDiagram](images/UndoSequenceDiagram.png)
+
+However, the `redo` command simply does the opposite and calls `Model#redoAddressBook()`, which causes the undone state to be reinstated.
+
+<div markdown="span" class="alert alert-info">:information_source: **Note:** If the current state is the latest 
+possible state, in this case tb2, Teletubbies will no longer be able to restore undone states, and the redo command will throw an exception. The undo command uses
```
Comment (skythefire) [inline docs/DeveloperGuide.md:456]: Same as above!

**Review by skythefire — APPROVED** [review]
LGTM, after fixing the typos (if they are indeed typos)!

File: docs/DeveloperGuide.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (tanqiya) [reply docs/DeveloperGuide.md:446]: it's intentional. referencing object name in the diagram

[the remark this answers is not in the document]
Comment (skythefire) [reply docs/DeveloperGuide.md:446]: I see

**Comment by skythefire** [comment]
For the Model Class Diagram, the multiplicity of Person to the optional fields ie. Email, Address, and Remark are 1 because in our implementation we give them default values?

**Comment by vigneshsankariyer1234567890** [comment]
> For the Model Class Diagram, the multiplicity of Person to the optional fields ie. Email, Address, and Remark are 1 because in our implementation we give them default values?

i wasnt thinking along those lines per se, more of the fact that Person and it's fields are immutable. This should allow the multiplicity to be 1

**Comment by gok99** [comment]
image

Should this be `c:HistoryCommand`?

**Comment by gok99** [comment]
Activity diagram conditions should be in box brackets (`[]`)

image

**Comment by vigneshsankariyer1234567890** [comment]
> image
> 
> Should this be `c:HistoryCommand`?
> 
> image
> 
> Perhaps could use `c:XYZCommand` with a note on what `XYZCommand` is, like in the Logic diagram:
> 
> image

good catch bro, i'll fix it asap before the next merge.

**Comment by vigneshsankariyer1234567890** [comment]
> Activity diagram conditions should be in box brackets (`[]`)
> 
> image

will rectify this as well

**Pull request merged**